nbresuse/prometheus.py

```diff
@@ -61,3 +69,44 @@ def apply_cpu_limit(self, cpu_metric_values) -> Optional[float]:
                 return self.config.cpu_limit
             else:
                 return 100.0 * cpu_metric_values["cpu_count"]
+            
+    @gen.coroutine
+    def kernel_metrics(self):
+        kernels = yield maybe_future(self._list_kernel_memory())
+        for kernel_id, data in kernels.items():
+            self.KERNEL_MEMORY_USAGE.labels(kernel_id=kernel_id).set(data['rss'])
+
+    @gen.coroutine
+    def _list_kernel_memory(self):
+        kernel_memory = dict()
+
+        session_manager = self.session_manager
+        sessions = yield maybe_future(session_manager.list_sessions())
+        kernel_processes = PrometheusHandler._kernel_processes()
+
+        def find_process(session_id):
+            for proc in kernel_processes:
+                cmd = proc.cmdline()
+                if cmd:
+                    last_arg = cmd[-1]
+                    if session_id in last_arg:
+                        return proc
+            return None
+
+        for session in sessions:
+            kernel = session['kernel']
+            kernel_id = kernel['id']
+            kernel_process = find_process(kernel_id)
+            if kernel_process:
+                kernel_memory[kernel_id] = kernel_process.memory_info()._asdict()
+
+        raise gen.Return(kernel_memory)
+
+    @staticmethod
+    def _kernel_processes():
+        for proc in process_iter():
+            try:
+                if 'ipykernel_launcher' in proc.cmdline():
```

Comment:
It looks like this could be tweaked to also handle other kernels than `ipykernel`?

Reply:
Yes that's correct.

Reply:
Hey @jtpio , are we good with PR or any more comments. Please do let me know.

Reply:
I wonder whether we could update this condition to also take other kernels into account. For example this is what the list of processes looks like with [xeus-python](https://github.com/jupyter-xeus/xeus-python) and ipykernel:

```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
jovyan         1  3.2  0.0 543436 72764 ?        Ssl  07:40   0:02 /srv/conda/envs/notebook/bin/python /srv/conda/envs/notebook/bin/jupyter-notebook --ip=0.0.0.0 --port=8888 --NotebookApp.base_url=/user/jupyterlab-debugger-669lwcpu/ --NotebookApp.token=RLdiB5_rTvucqwNKL
jovyan        87  0.4  0.0 351976 23264 ?        Ssl  07:41   0:00 /srv/conda/envs/notebook/bin/xpython -f /home/jovyan/.local/share/jupyter/runtime/kernel-0551efcc-dca9-4d4a-8e8d-ea0cd6283426.json
jovyan       119  5.2  0.0 453612 48156 ?        Ssl  07:41   0:00 /srv/conda/envs/notebook/bin/python -m ipykernel_launcher -f /home/jovyan/.local/share/jupyter/runtime/kernel-70ec4260-e6b3-44de-aece-ab101da467a1.json
```

Reply:
Maybe a match against something like `kernel-(\w+-\w+-\w+-\w+-\w+).json`.

Reply:
This might capture other kernels, but I am not so sure if this regex will actually capture just the kernel processes. What if I just opened a file or any other non-jupyter process using this kind of regex?

Reply:
Right. It's also possible to connect to the same kernel from another client using this connection file.

But wouldn't there be a similar issue with the `if 'ipykernel_launcher' in proc.cmdline()` check?

For example one could monitor the list of kernels by running this command: `top -c | grep ipykernel_launcher`

This will start a new process that contains `ipykernel_launcher` in the command.

Reply:
If you see line 87, we are filtering processes using `kernel_id`. Similarly, I think we can capture other kernels info given if we can identify the processes and later we apply above filter using `kernel_id`. What do you think?

Reply:
So it looks like that piece of logic does rely on finding the kernel process using the name of the connection file?

https://github.com/yuvipanda/nbresuse/pull/41/files#diff-a91d5ce40a78500dc1186d016d12288bR87-R94

Basically finding the kernel id in `/home/jovyan/.local/share/jupyter/runtime/kernel-0551efcc-dca9-4d4a-8e8d-ea0cd6283426.json`

Reply:
That's Correct.

Reply:
> What if I just opened a file or any other non-jupyter process using this kind of regex?

So we'll probably face a similar issue as this one?

Reply:
(although we could assume this is unlikely going to be the case for now)

Reply:
I think we won't since first filter is `ipykernel_launcher `

Reply:
If the first filter was based on `kernel_id` then we might face that issue.

Reply:
Should the two filters be kept separate?

Reply:
One way to support all types of kernels could be to do the match based on the `argv` field from the kernelspec: https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs

Reply:
I just kept separate thinking only about python kernels. I am open to suggestions to make this generic.

Reply:
This looks good. `argv` has more fields to match and filter.

Reply:
Probably with `nbapp.kernel_spec_manager`

Reply:
Yeah I am making this change.

Reply:
Awesome, thanks!

Reply:
@jtpio For xeus-python, the process has `xpython_launcher`. I can use this to filter similar to ipython.

`/bin/python -m xpython_launcher -f /Users/gsaibhar/Library/Jupyter/runtime/kernel-80b1a6ec-47df-4467-b093-54f49b1f623c.json`

Reply:
`argv` doesn't have more info than what we see in the process. I am not sure how we can support all kernels at this point of time.

Reply:
> I can use this to filter similar to ipython.
> 
> `/bin/python -m xpython_launcher -f /Users/gsaibhar/Library/Jupyter/runtime/kernel-80b1a6ec-47df-4467-b093-54f49b1f623c.json`

`xeus-python` can also be installed with `conda`. In that case the command to start the kernel doesn't contain `xpython_launcher`:

```
/srv/conda/envs/notebook/bin/xpython -f /home/jovyan/.local/share/jupyter/runtime/kernel-0551efcc-dca9-4d4a-8e8d-ea0cd6283426.json
```

Reply:
If we can retrieve the list of kernel specs, we should be able to construct a list of argvs to designate what the commands of the kernel processes should look like. For example:

```
[
  ["/srv/conda/envs/notebook/bin/python", "-m", "ipython_launcher", "{connection_file}"],
  ["/srv/conda/envs/notebook/bin/python", "-m", "xpython_launcher", "{connection_file}"]
  ["/srv/conda/envs/notebook/bin/xpython", "-f", "{connection_file}"]
``` 

Then in the filter, we could match each of the items above against `proc.cmdline()` (also using the kernel id).

Reply:
I am now able to capture using `argv` and able to get `xpython` as well in metrics.

Reply:
Nice!